bluemira/equilibria/solve_new.py

```diff
@@ -0,0 +1,635 @@
+
+        Returns
+        -------
+        converged: bool
+            True if the convergence criterion is met, else False.
+        """
+        conv = np.linalg.norm(j_old - j_new)
+        if print_status:
+            # Format convergence
+            bluemira_print_flush(
+                f"EQUILIBRIA G-S iter {i}: ||J_phi_old-J_phi|| convergence: {conv:e}"
+            )
+        return self.check_converged(conv)
+
+
+class PicardIterator:
+    """
+    A Picard iterative solver.
+
+    Child classes must provide a __call__ method which carries out the
+    iteration process(es)
+
+    Parameters
+    ----------
+    eq: Equilibrium object
+        The equilibrium to solve for
+    profiles: Profile object
+        The plasma profile object to solve with
+    constraints: Constraint object
+        The constraint to solve for
+    optimiser: EquilibriumOptimiser object
+        The optimiser to use
+    convergence: ConvergenceCriterion
+        The convergence criterion to use
+    relaxation: float
+        The relaxation parameter to use between iterations
+    maxiter: int
+        The maximum number of iterations
+    plot: bool
+        Whether or not to plot
+    gif: bool
+        Whether or not to make a GIF
+    figure_folder: str (default = None)
+        The path where figures will be saved. If the input value is None (e.g. default)
+        then this will be reinterpreted as the path data/plots/equilibria under the
+        bluemira root folder, if that path is available.
+    """
+
+    def __init__(
+        self,
+        eq,
+        profiles,
+        optimisation_problem,
+        convergence=DudsonConvergence(),
+        I_not_dI: bool = False,
+        fixed_coils: bool = False,
+        relaxation: int = 0,
+        maxiter: int = 30,
+        plot=True,
+        gif=False,
+        figure_folder=None,
+        plot_name="default_0",
+    ):
+        self.eq = eq
+        self.coilset = self.eq.coilset
+        self.profiles = profiles
+        self.opt_prob = optimisation_problem
+        if isinstance(convergence, ConvergenceCriterion):
+            self.convergence = convergence
+        else:
+            raise ValueError(
+                "Optimiser convergence specification must be a sub-class of ConvergenceCriterion."
+            )
+        self.I_not_dI = I_not_dI
+        self.fixed_coils = fixed_coils
+
+        self.relaxation = relaxation
+        self.maxiter = maxiter
+        self.plot_flag = plot
+        self.gif_flag = gif
+        if figure_folder is None:
+            figure_folder = try_get_bluemira_path(
+                "plots/equilibria", subfolder="data", allow_missing=not self.gif_flag
+            )
+        self.figure_folder = figure_folder
+        self.store = []
+        self.i = 0
+        if self.plot_flag:
+            self.pname = plot_name
+            self.f, self.ax = plt.subplots()
+
+    def _optimise_currents(self):
```

Comment:
Could possibly rename this to `_optimise_coilset,` as `coilset` attributes other than just currents may be optimised